demos/source/aws_iot_demo_jobs.c

```diff
@@ -0,0 +1,591 @@
+
+            status = err == IOT_MQTT_SUCCESS;
+        }
+        else
+        {
+            IotLogInfo( "Failed to execute publish." );
+        }
+    }
+
+    /*
+     * The "print" command prints the message to console (with quotes if it's a JSON string).
+     *
+     * An example JSON doc:
+     * ```
+     * {"action":"print","message":"Hello world!"}
+     * ```
+     *
+     * This will print "Hello world!" in the device log.
+     *
+     * If message is missing from the JSON job document, the execution will fail.
+     */
+    else if( COMPARE_COMMAND( "print" ) )
+    {
+        if( _getMessage( &pMessage, &messageLength ) )
+        {
+            IotLogInfo( "Executing Print: %.*s", messageLength, pMessage );
+            status = true;
+        }
+        else
+        {
+            IotLogInfo( "Failed to execute print." );
+        }
+    }
+    else
+    {
+        IotLogError( "Unrecognized command: %.*s", commandLength, command );
+    }
+
+    return status;
+
+#undef COMPARE_COMMAND
+}
+
+/**
+ * @brief Get and update the job.
+ */
+static int32_t _executeDemo( IotMqttConnection_t const mqttConnection,
+                             const char * pThingName,
+                             size_t thingNameLength )
+{
+    bool success = false;
+    AwsIotJobsError_t err = AWS_IOT_JOBS_SUCCESS;
+    AwsIotJobState_t result = AWS_IOT_JOB_STATE_FAILED;
+    AwsIotJobsResponse_t resp = AWS_IOT_JOBS_RESPONSE_INITIALIZER;
+    AwsIotJobsRequestInfo_t req = AWS_IOT_JOBS_REQUEST_INFO_INITIALIZER;
+    AwsIotJobsUpdateInfo_t updateInfo = AWS_IOT_JOBS_UPDATE_INFO_INITIALIZER;
+    const char * pAction = NULL;
+    size_t actionLength = 0;
+
+    req.mqttConnection = mqttConnection;
+    req.pThingName = pThingName;
+    req.thingNameLength = thingNameLength;
+    req.pJobId = _pJobId + 1;
+    req.jobIdLength = _JobIdLength - 2;
+    req.mallocResponse = malloc;
+
+    /* Check that a job notification came in. */
+    if( _pJobId[ 0 ] != '\0' )
+    {
+        /* Update Job to IN_PENDING. */
+        updateInfo.newStatus = AWS_IOT_JOB_STATE_IN_PROGRESS;
+
+        err = AwsIotJobs_UpdateSync( &req, &updateInfo, 0, TIMEOUT_MS, &resp );
+
+        if( err != AWS_IOT_JOBS_SUCCESS )
+        {
+            IotLogError( "Update Failed: %s", AwsIotJobs_strerror( err ) );
+            success = false;
+        }
+        else
+        {
+            success = true;
+        }
+    }
+
+    /* Parse out the action. */
+    if( success )
+    {
+        success = _getAction( &pAction, &actionLength );
+
+        if( !success )
+        {
+            IotLogError( "Failed to find %s Key.", JOB_ACTION_KEY );
+        }
+    }
+
+    /* Process the action. */
+    if( success )
+    {
+        /* Execute on the action without quotes. */
+        if( _executeCommand( pAction + 1, actionLength - 1, mqttConnection ) )
+        {
+            result = AWS_IOT_JOB_STATE_SUCCEEDED;
+        }
```

Comment:
If _executeCommand fails, you should still send the job update with status AWS_IOT_JOB_STATE_FAILED.

Reply:
It is based on initialization.